src/Shared/Recruit.Vacancies.Client/Infrastructure/Services/Projections/PublishedVacancyProjectionService.cs

```diff
@@ -0,0 +1,78 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Esfa.Recruit.Vacancies.Client.Domain.Entities;
+using Esfa.Recruit.Vacancies.Client.Domain.Repositories;
+using Esfa.Recruit.Vacancies.Client.Infrastructure.Extensions;
+using Esfa.Recruit.Vacancies.Client.Infrastructure.QueryStore;
+using Esfa.Recruit.Vacancies.Client.Infrastructure.QueryStore.Projections.Vacancy;
+using Esfa.Recruit.Vacancies.Client.Infrastructure.ReferenceData;
+using Esfa.Recruit.Vacancies.Client.Infrastructure.ReferenceData.ApprenticeshipProgrammes;
+using Microsoft.Extensions.Logging;
+
+namespace Esfa.Recruit.Vacancies.Client.Infrastructure.Services.Projections
+{
+    public class PublishedVacancyProjectionService : IPublishedVacancyProjectionService
+    {
+        private readonly ILogger<PublishedVacancyProjectionService> _logger;
+        private readonly IQueryStoreWriter _queryStoreWriter;
+        private readonly IVacancyRepository _repository;
+        private readonly IReferenceDataReader _referenceDataReader;
+
+        public PublishedVacancyProjectionService(
+                                    ILogger<PublishedVacancyProjectionService> logger, 
+                                    IQueryStoreWriter queryStoreWriter, 
+                                    IVacancyRepository repository,
+                                    IReferenceDataReader referenceDataReader)
+        {
+            _logger = logger;
+            _queryStoreWriter = queryStoreWriter;
+            _repository = repository;
+            _referenceDataReader = referenceDataReader;
+        }
+
+        public async Task ReGeneratePublishedVacanciesAsync()
+        {            
+            var liveVacanciesTask = _repository.GetVacanciesByStatusAsync(VacancyStatus.Live);
+            var closedVacanciesTask = _repository.GetVacanciesByStatusAsync(VacancyStatus.Closed);
+            var programmesTask = _referenceDataReader.GetReferenceData<ApprenticeshipProgrammes>();
+
+            await Task.WhenAll(liveVacanciesTask, programmesTask, closedVacanciesTask);
+
+            var liveVacancies = liveVacanciesTask.Result.ToList();
+            var programmesData = programmesTask.Result.Data;
+            var closedVacancies = closedVacanciesTask.Result.ToList();
+
+            var regenerateLiveVacanciesViewsTask = RegenerateLiveVacancies(liveVacancies, programmesData);
+            var regenerateClosedVacanciesViewsTask = RegenerateClosedVacancies(closedVacancies, programmesData);
+
+            await Task.WhenAll(regenerateClosedVacanciesViewsTask, regenerateLiveVacanciesViewsTask);
+        }
+
+        private Task RegenerateLiveVacancies(List<Vacancy> liveVacancies, List<ApprenticeshipProgramme> programmes)
```

Comment:
Don't think you need to pass the args in as List, could be IEnumerable.

Reply:
I am using list to avoid double iteration as I need to log the count.